Homeworks/Homework_5/part_4.py

```diff
@@ -0,0 +1,130 @@
+    user_with_discounts = {user_id: random.choice(credits_discount) for user_id in user_ids}
+    logger.info("Discount assignment completed")
+    return user_with_discounts
+
+
+@establish_db_connection
+def get_debtors_full_name(cursor):
+    """
+    Retrieves the full names of users with negative account balances.
+
+    :return: list: A list of full names of users with negative account balances.
+    """
+    logger.info("Retrieving full names of users with negative account balances")
+    debtors_id = [i[0] for i in cursor.execute('SELECT User_id FROM Account WHERE Amount < 0')]
+    return [' '.join(list(cursor.execute(f'SELECT Name, Surname FROM User WHERE Id = {debtor_id}'))[0])
+            for debtor_id in debtors_id]
+
+
+@establish_db_connection
+def get_bank_with_the_biggest_capital(cursor):
+    """
+    Retrieves the bank with the highest capital in converted currency.
+
+    :return: tuple: A tuple containing a dictionary of bank capitals and the name of the bank
+                    with the highest capital.
+    """
+    logger.info("Retrieving bank with the highest capital")
+    currency = get_currency()
+    banks_capital = {}
+
+    cursor.execute('SELECT Bank_id, Currency, Amount FROM Account')
+    rows = cursor.fetchall()
+
+    for row in rows:
+        bank_id, account_currency, account_amount = row
+        converted_amount = account_amount * currency[account_currency]
+        bank_name = list(cursor.execute(f'SELECT name FROM Bank WHERE id = {bank_id}'))[0]
+        banks_capital.setdefault(bank_name, 0)
+        banks_capital[bank_name] += converted_amount
+    logger.info("Capital retrieval completed")
+    return max(banks_capital)[0]
+
+
+@establish_db_connection
+def get_bank_with_the_oldest_client(cursor):
+    """
+    Retrieves the name of the bank with the oldest client.
+
+    :return: str: The name of the bank with the oldest client.
+    """
+    logger.info("Retrieving bank with the oldest client")
+    clients_birth_day = list(cursor.execute(f'SELECT id, Birth_day FROM User '))
+    the_oldest_user = min(clients_birth_day, key=lambda client: datetime.strptime(client[1], '%d-%m-%Y'))
+    cursor.execute(f'SELECT Bank_id FROM Account WHERE User_id = {the_oldest_user[0]}')
+    the_oldest_user_bank_id = cursor.fetchone()[0]
+    cursor.execute(f'SELECT name FROM Bank WHERE id = {the_oldest_user_bank_id}')
+    return cursor.fetchone()[0]
+
+
+@establish_db_connection
+def get_bank_with_highest_unique_outbound_users(cursor):
+    """
+    Retrieves the bank with the highest number of unique users with outbound transactions.
+
+    :return: tuple: A tuple containing the bank name and the count of unique outbound users.
+    """
+    logger.info("Retrieving bank with the highest unique outbound users")
+    transactions_data = set([i for i in list(cursor.execute('SELECT Bank_sender_name, Account_sender_id '
+                                                            'FROM Transactions'))])
```

Comment:
why do you create list if you immediately convert it to set()
and why do u need a comprehension here, at all?))